### PR TITLE
Refactor: Dependency injection pattern

### LINKED ISSUE #20: Refactor to dependency injection
Pretty self explanatory. We want a "core" library that can be used in a variety of created plugins.

We eventually want the same API to work with different "table implementations" and different "schema implementations".

Discussion of this is covered somewhat in #15 .

### PR DESCRIPTION
**Pull request by wesselhuising**
This MR includes the work that has been discussed together with @xaviernogueira. We decided to refactor the project with the goal to become more flexible in its future usage, by using the dependency injection pattern.

The main class is renamed to `Pandantic` and would need a Schema of the type` Pydantic.BaseModel` to be successfully instantiated. 

Everything in this PR is API breaking change and should be tagged as a major bump for the next release.

### CODE REVIEW

**Comment by xaviernogueira** [comment]
@wesselhuising sick! I'll review this shortly (if you think it's ready).

Also mind briefly reviewing/merging my other PR for now. It is non-breaking.

**Comment by wesselhuising** [comment]
I am fixing up the last failing test (pylint) and you can start shooting!

edit: @xaviernogueira go for it, my dev env was not working 100% for pylint for some reason so it has been a trial and error for the pylint part when it comes to commits (so def need to squash lol). Still not working properly but we'll figure that one out later when the code is finished.

File: src/pandantic/basemodel.py
Comment (xaviernogueira) [inline src/pandantic/basemodel.py:3]: Not a comment on an addition, but why is this line here? My linter says it is used, but i don't see where. More a curiosity question.

File: src/pandantic/types.py
Comment (xaviernogueira) [inline src/pandantic/types.py]: For now I think "DataFrame" makes sense, but perhaps it's worth considering "Table". Some libraries like `pyarrow` call it a "Table", as well as the database context. 

I could see a world where we define an interface between some sort of Table object for validation, where this could be used beyond in-memory dataframes (i.e., CSV/Parquet files directly, dbase tables, etc.), but that is out-of-scope!

File: src/pandantic/validators/pandas.py
Comment (xaviernogueira) [inline src/pandantic/validators/pandas.py:18]: I like the sub-classing of `BaseValidator`! Nice idea

File: src/pandantic/validators/baseclass.py
Comment (xaviernogueira) [inline src/pandantic/validators/baseclass.py:1]: Maybe rename the file `base.py`? I feel like `validators.baseclass.BaseValidator` is a bit verbose.

File: src/pandantic/validators/baseclass.py
Comment (xaviernogueira) [inline src/pandantic/validators/baseclass.py]: What do you think about having the function return a bool? I feel semantically that makes the most sense ("validate" -> is it valid -> yes/no -> bool). Also I think without that an Any-> Any base class is somewhat meaningless.

We can more meaningfully build upon the BaseValidator subless (i.e., `pandas` plugin, other features, etc.) with a well defines validation behavior.

File: src/pandantic/validators/pandas.py
Comment (xaviernogueira) [inline src/pandantic/validators/pandas.py:25]: Kinda a fun idea, but what about providing type safety for the functions using [`pydantic` decorators](https://docs.pydantic.dev/latest/concepts/validation_decorator/)? I've never used them before, but it could be cool as it double-down on the whole "pydantic" ecosystem/pattern you see in `FastAPI` and other

File: src/pandantic/validators/pandas.py
Comment (xaviernogueira) [inline src/pandantic/validators/pandas.py]: Let's see if we can avoid doing any copying. With large dataframes that could pose a memory risk, especially if people want to deploy this in async services where it could be running in parallel with other dataframes.

File: src/pandantic/validators/pandas.py
Comment (xaviernogueira) [inline src/pandantic/validators/pandas.py]: Mind catching the specific exception `PydanticValidationError`, so we can handle other (unexpected) errors differently.

File: tests/test_custom_validator_context_str_list.py
Comment (xaviernogueira) [inline tests/test_custom_validator_context_str_list.py:1]: The custom validator tests should probably get combined, as the test directory will get huge at this level of granularity per `.py` file.

**Review by xaviernogueira — CHANGES_REQUESTED** [review]
So I added a couple comments with things I think could be changes. 

But a bigger picture change idea is this:
* Currently you have `CoreValidator` aka `Pandantic`, therefore a user would for instantiate `Pandantic(schema)` and run `.validate`.
* This requires users to needlessly re-instantiate many times in some cases, or do two lines of code instead of one.
* What if the behavior is `pandantic.validate(df, schema)`, and the one function has a decorator that replicates the `_get_implementation()` logic, and injects the correct class?

I also need to update the pandas plugin code, so I will make a branch off this, and add the decorator pattern (on top of the existing pattern) to demo. Then we can discuss.

Thanks for the work! And sorry I took so long to get to this.

**Comment by xaviernogueira** [comment]
Oh another comment, but it's missing support for the "filter" behavior that previously existed. I am feeling motivated so if you don't mind I will add it in my PR!

File: src/pandantic/validators/pandas.py
Comment (xaviernogueira) [inline src/pandantic/validators/pandas.py]: We should have the Queue as an optional argument. One might want to abstract the multiprocessing aspect, or run an async Queue or something like that. The default should remain as is.

File: src/pandantic/validators/baseclass.py
Comment (xaviernogueira) [reply src/pandantic/validators/baseclass.py]: That said I also see why you may want to support the previous behavior. I personally am in favor of validate and filter being separate functions as well.

File: src/pandantic/validators/pandas.py
Comment (xaviernogueira) [reply src/pandantic/validators/pandas.py:25]: This actually only works with more primitive input types. Nevermind!

File: src/pandantic/validators/baseclass.py
Comment (xaviernogueira) [reply src/pandantic/validators/baseclass.py]: This does not need to be addressed in this PR, more food for though. Feel free to resolve.

**Comment by xaviernogueira** [comment]
@wesselhuising okay so I fixed the pandas plugin stuff to match you changes! Feel free to address my comments whenever you can, and if you want I am happy to discuss further.

My next step will be a branch off of this to simplify the user pattern (i.e., `pydantic.valiadate`, pydantic.filter`) and will try to add a the `iterrow` function.

**Regarding my requested changes**, I wrote a lot a bit stream of consciousness, so this is the key changes I think should happen before this is merged:
* Make `Queue` passed in as an optional argument, with a new `Queue` made if None.
* Fix the `errors=filter` logic for when `n_jobs > 1`. 
* Catch the `PydanticValidationError` specifically, catch other errors separately.
* Rename `baseclass` -> `base` to reduce path verbosity.
* Bonus (if possible/easy): Remove the dataframe copying. This might be a later task, as copy free multiprocessing can be tricky.

Again thanks for the hard work! I am handling the tests consolidation as I am adding new tests.

File: src/pandantic/validators/pandas.py
Comment (xaviernogueira) [inline src/pandantic/validators/pandas.py:83]: **This logic is incorrect!** The invalid chunks get messaged back out as invalid rows, which then will remove too many rows if `errors="filter"`.

The easiest fix is probably changing the `_validate_chunks` function to return errors row indices, then re-combing these lists.

File: src/pandantic/basemodel.py
Comment (wesselhuising) [reply src/pandantic/basemodel.py:3]: What are you using for linting?

File: src/pandantic/types.py
Comment (wesselhuising) [reply src/pandantic/types.py]: I do like the Table naming suggestion. Shall I change it while we are at it currently?

Comment (wesselhuising) [reply src/pandantic/types.py]: Changes the name to `TableTypes`.

File: src/pandantic/validators/baseclass.py
Comment (wesselhuising) [reply src/pandantic/validators/baseclass.py:1]: Done.

File: src/pandantic/validators/baseclass.py
Comment (wesselhuising) [reply src/pandantic/validators/baseclass.py]: I am in favor of changing this in another PR, as the scope for this PR is more or less about the dependency injection pattern and want to avoid the PR becoming very broad scoped.

Comment (wesselhuising) [reply src/pandantic/validators/baseclass.py]: I do agree though, might want to create a ticket for this increment.

File: src/pandantic/validators/pandas.py
Comment (wesselhuising) [reply src/pandantic/validators/pandas.py]: Agree this is a waste of resources but I am not sure how to move forward without copying and not mutation the original dataframe. Maybe with the new future methods (validate & filter) we could look into this challenge, maybe creating an index table over copying the dataframe?

File: src/pandantic/validators/pandas.py
Comment (wesselhuising) [reply src/pandantic/validators/pandas.py]: Done, good catch.

File: tests/test_custom_validator_context_str_list.py
Comment (wesselhuising) [reply tests/test_custom_validator_context_str_list.py:1]: 100%, do you feel that it should be part of this PR or can we move it into a new one just focussing on refactoring the testing part? I feel more for the latter option.

File: src/pandantic/validators/pandas.py
Comment (wesselhuising) [reply src/pandantic/validators/pandas.py]: Moved it to the arguments as an optional Queue object. Will only be defaulted to Queue if nothing is specified in the scope of the multiprocessing run in case of the amount of jobs is not 1. 

I was wondering, should I change the typing to Any as we want to allow multiple Queue classes from multiple packages?

File: src/pandantic/validators/pandas.py
Comment (wesselhuising) [reply src/pandantic/validators/pandas.py:83]: Are you sure, only the validation error rows are being put into the queue and whenever the chunk is finished, a None is put onto to the Queue indicating the batch is done. I did change the name of the function though.

**Comment by wesselhuising** [comment]
I'll wait for your comments and resolves and fix the CI tests before moving on. Cheers.

File: src/pandantic/basemodel.py
Comment (xaviernogueira) [reply src/pandantic/basemodel.py:3]: I use `pyright`, if it looks fine on `mypy` or whatever you are using I say this doesn't matter. Btw what is this for?

File: src/pandantic/validators/pandas.py
Comment (xaviernogueira) [reply src/pandantic/validators/pandas.py]: We can avoid mutating the original dataframe by not adding columns to it, but just reading values from it. Since we convert everything into a dict at one point, we might as well incur that cost easy and just use the dicts such that we only copy once (the dicts), versus technically three times (df1, df2, dict version).

**I just went ahead and implemented this!** Small tweak where we use `df.to_dict("Index")` which produces a `dict[index, dict[column, value]`. Such that we can iterate over  the .items() of the dict and get the index (without making a new column) and the data for validation.

File: src/pandantic/validators/pandas.py
Comment (xaviernogueira) [reply src/pandantic/validators/pandas.py:83]: Ah I see, yes that makes sense, I was misunderstanding the use of the queue.

**Review by xaviernogueira — APPROVED** [review]

**Comment by xaviernogueira** [comment]
@wesselhuising so I implemented a copy free version and all the tests are passing again (also had to switch `ValidationError` -> `ValueError` in the custom validator tests)! 

Take a look when you get the chance! I'm feeling pretty good about it all.

Btw for some reason it's not letting me request a review from you.

**Pull request merged**
